File: common/src/main/java/net/impactdev/gts/common/messaging/messages/listings/buyitnow/purchase/BINPurchaseMessage.java
package net.impactdev.gts.common.messaging.messages.listings.buyitnow.purchase;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import net.impactdev.gts.api.messaging.message.errors.ErrorCode;
import net.impactdev.gts.api.messaging.message.type.listings.BuyItNowMessage;
import net.impactdev.gts.common.messaging.GTSMessagingService;
import net.impactdev.gts.api.messaging.message.errors.ErrorCodes;
import net.impactdev.gts.common.messaging.messages.AbstractMessage;
import net.impactdev.gts.common.plugin.GTSPlugin;
import net.impactdev.impactor.api.json.factory.JObject;
import net.impactdev.impactor.api.utilities.printing.PrettyPrinter;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public abstract class BINPurchaseMessage extends AbstractMessage implements BuyItNowMessage.Purchase {

    protected final UUID listing;
    protected final UUID actor;

    public BINPurchaseMessage(UUID id, UUID listing, UUID actor) {
        super(id);
        this.listing = listing;
        this.actor = actor;
    }

    @Override
    public UUID getListingID() {
        return this.listing;
    }

    @Override
    public UUID getActor() {
        return this.actor;
    }

    public static class Request extends BINPurchaseMessage implements Purchase.Request {

        public static final String TYPE = "BIN/Purchase/Request";

        public static BINPurchaseMessage.Request decode(@Nullable JsonElement content, UUID id) {
            if(content == null) {
                throw new IllegalStateException("Raw JSON data was null");
            }

            JsonObject raw = content.getAsJsonObject();

            UUID listing = Optional.ofNullable(raw.get("listing"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate listing ID"));
            UUID actor = Optional.ofNullable(raw.get("actor"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate actor ID"));

            return new BINPurchaseMessage.Request(id, listing, actor);
        }

        public Request(UUID id, UUID listing, UUID actor) {
            super(id, listing, actor);
        }

        @Override
        public @NonNull String asEncodedString() {
            return GTSMessagingService.encodeMessageAsString(
                    TYPE,
                    this.getID(),
                    new JObject()
                            .add("listing", this.listing.toString())
                            .add("actor", this.actor.toString())
                            .toJson()
            );
        }


        @Override
        public CompletableFuture<Purchase.Response> respond() {
            return GTSPlugin.instance().storage().processPurchase(this);
        }

        @Override
        public void print(PrettyPrinter printer) {
            printer.kv("Request ID", this.getID())
                    .kv("Listing ID", this.getListingID())
                    .kv("Actor", this.getActor());
        }
    }

    public static class Response extends BINPurchaseMessage implements Purchase.Response {

        public static final String TYPE = "BIN/Purchase/Response";

        public static BINPurchaseMessage.Response decode(@Nullable JsonElement content, UUID id) {
            if (content == null) {
                throw new IllegalStateException("Raw JSON data was null");
            }

            JsonObject raw = content.getAsJsonObject();

            UUID listing = Optional.ofNullable(raw.get("listing"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate listing ID"));
            UUID actor = Optional.ofNullable(raw.get("actor"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate actor ID"));
            UUID request = Optional.ofNullable(raw.get("request"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate request ID"));
            UUID seller = Optional.ofNullable(raw.get("seller"))
                    .map(x -> UUID.fromString(x.getAsString()))
                    .orElseThrow(() -> new IllegalStateException("Unable to locate seller ID"));
            boolean successful = Optional.ofNullable(raw.get("successful"))
                    .map(JsonElement::getAsBoolean)
                    .orElseThrow(() -> new IllegalStateException("Unable to locate successful status marker"));
            ErrorCode error = Optional.ofNullable(raw.get("error"))
                    .map(x -> ErrorCodes.get(x.getAsInt()))
                    .orElse(null);

            return new BINPurchaseMessage.Response(id, request, listing, actor, seller, successful, error);
        }

        private final UUID request;
        private final UUID seller;
        private final boolean successful;
        private long time;

        private final ErrorCode error;

        public Response(UUID id, UUID request, UUID listing, UUID actor, UUID seller, boolean successful, ErrorCode error) {
            super(id, listing, actor);
            this.request = request;
            this.seller = seller;
            this.successful = successful;
            this.error = error;
        }

        @Override
        public @NonNull String asEncodedString() {
            return GTSMessagingService.encodeMessageAsString(
                    TYPE,
                    this.getID(),
                    new JObject()
                            .add("request", this.request.toString())
                            .add("listing", this.listing.toString())
                            .add("actor", this.actor.toString())
                            .add("seller", this.seller.toString())
                            .add("successful", this.successful)
                            .consume(o -> this.getErrorCode().ifPresent(e -> o.add("error", e.ordinal())))
                            .toJson()
            );
        }

        @Override
        public UUID getRequestID() {
            return this.request;
        }

        @Override
        public long getResponseTime() {
            return this.time;
        }

        @Override
        public void setResponseTime(long millis) {
            this.time = millis;
        }

        @Override
        public boolean wasSuccessful() {
            return this.successful;
        }

        @Override
        public Optional<ErrorCode> getErrorCode() {
            return Optional.ofNullable(this.error);
        }

        @Override
        public void print(PrettyPrinter printer) {
            printer.kv("Response ID", this.getID())
                    .kv("Request ID", this.getRequestID())
                    .kv("Listing ID", this.getListingID())
                    .kv("Actor", this.getActor())
                    .kv("Seller", this.getSeller());
        }

        @Override
        public UUID getSeller() {
            return this.seller;
        }
    }

}

File: common/src/main/kotlin/net/impactdev/gts/common/messaging/messages/listings/buyitnow/purchase/BINPurchaseMessage.kt
package net.impactdev.gts.common.messaging.messages.listings.buyitnow.purchase

import com.google.gson.JsonElement
import net.impactdev.gts.api.messaging.message.errors.ErrorCode
import net.impactdev.gts.api.messaging.message.errors.ErrorCodes
import net.impactdev.gts.api.messaging.message.type.listings.BuyItNowMessage.Purchase
import net.impactdev.gts.api.util.PrettyPrinter
import net.impactdev.gts.common.messaging.GTSMessagingService
import net.impactdev.gts.common.messaging.messages.AbstractMessage
import net.impactdev.gts.common.plugin.GTSPlugin
import net.impactdev.impactor.api.json.factory.JObject
import java.util.*
import java.util.concurrent.CompletableFuture

abstract class BINPurchaseMessage(id: UUID, override val listingID: UUID?, override val actor: UUID?) :
    AbstractMessage(id), Purchase {

    class Request(id: UUID, listing: UUID?, actor: UUID?) : BINPurchaseMessage(id, listing, actor), Purchase.Request {
        override fun asEncodedString(): String {
            return GTSMessagingService.Companion.encodeMessageAsString(
                TYPE,
                id,
                JObject()
                    .add("listing", listingID.toString())
                    .add("actor", actor.toString())
                    .toJson()
            )
        }

        override fun respond(): CompletableFuture<Purchase.Response?>? {
            return GTSPlugin.Companion.getInstance().getStorage().processPurchase(this)
        }

        override fun print(printer: PrettyPrinter) {
            printer.kv("Request ID", id)
                .kv("Listing ID", listingID)
                .kv("Actor", actor)
        }

        companion object {
            const val TYPE = "BIN/Purchase/Request"
            @kotlin.jvm.JvmStatic
            fun decode(content: JsonElement?, id: UUID): Request {
                checkNotNull(content) { "Raw JSON data was null" }
                val raw = content.asJsonObject
                val listing = Optional.ofNullable(raw["listing"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate listing ID") }
                val actor = Optional.ofNullable(raw["actor"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate actor ID") }
                return Request(id, listing, actor)
            }
        }
    }

    class Response(
        id: UUID,
        override val requestID: UUID,
        listing: UUID?,
        actor: UUID?,
        override val seller: UUID?,
        private val successful: Boolean,
        private val error: ErrorCode?
    ) : BINPurchaseMessage(id, listing, actor), Purchase.Response {
        override var responseTime: Long = 0
        override fun asEncodedString(): String {
            return GTSMessagingService.Companion.encodeMessageAsString(
                TYPE,
                id,
                JObject()
                    .add("request", requestID.toString())
                    .add("listing", listingID.toString())
                    .add("actor", actor.toString())
                    .add("seller", seller.toString())
                    .add("successful", successful)
                    .consume { o: JObject -> errorCode.ifPresent { e: ErrorCode -> o.add("error", e.ordinal()) } }
                    .toJson()
            )
        }

        override fun wasSuccessful(): Boolean {
            return successful
        }

        override val errorCode: Optional<ErrorCode>
            get() = Optional.ofNullable(error)

        override fun print(printer: PrettyPrinter) {
            printer.kv("Response ID", id)
                .kv("Request ID", requestID)
                .kv("Listing ID", listingID)
                .kv("Actor", actor)
                .kv("Seller", seller)
        }

        companion object {
            const val TYPE = "BIN/Purchase/Response"
            @kotlin.jvm.JvmStatic
            fun decode(content: JsonElement?, id: UUID): Response {
                checkNotNull(content) { "Raw JSON data was null" }
                val raw = content.asJsonObject
                val listing = Optional.ofNullable(raw["listing"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate listing ID") }
                val actor = Optional.ofNullable(raw["actor"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate actor ID") }
                val request = Optional.ofNullable(raw["request"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate request ID") }
                val seller = Optional.ofNullable(raw["seller"])
                    .map { x: JsonElement -> UUID.fromString(x.asString) }
                    .orElseThrow { IllegalStateException("Unable to locate seller ID") }
                val successful = Optional.ofNullable(raw["successful"])
                    .map { obj: JsonElement -> obj.asBoolean }
                    .orElseThrow { IllegalStateException("Unable to locate successful status marker") }
                val error = Optional.ofNullable(raw["error"])
                    .map { x: JsonElement -> ErrorCodes[x.asInt] }
                    .orElse(null)
                return Response(id, request, listing, actor, seller, successful, error)
            }
        }
    }
}